### PR TITLE
[WIP] Add MSD radix sort

### PR DESCRIPTION
**Pull request by kvinty**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):
Implemented MSD radix sort (based on kxsort), and partial sorting

### CODE REVIEW

File: dbms/src/AggregateFunctions/QuantileTDigest.h [hunk 1]
```diff
@@ -136,7 +136,7 @@ class QuantileTDigest
     {
         if (unmerged > 0)
         {
-            RadixSort<RadixSortTraits>::execute(summary.data(), summary.size());
+            RadixSort<RadixSortTraits>::executeLsd(summary.data(), summary.size());
```
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/QuantileTDigest.h]: Abbreviations are always in all caps, example: HTML, not Html, JSON, not Json.

File: dbms/src/Common/RadixSort.h [hunk 2]
```diff
@@ -43,61 +43,36 @@ struct RadixSortMallocAllocator
...
-};
-
-template <typename KeyBits>
-struct RadixSortIdentityTransform
-{
-    static constexpr bool transform_is_simple = true;
 
-    static KeyBits forward(KeyBits x)     { return x; }
-    static KeyBits backward(KeyBits x)    { return x; }
+    static bool compare(TElement x, TElement y)
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: It is named `less`, not `compare`.

Comment (alexey-milovidov) [reply dbms/src/Common/RadixSort.h]: Why we compare Elements, should we compare Keys instead?

File: dbms/src/Common/RadixSort.h [hunk 3]
```diff
@@ -163,6 +180,8 @@ struct RadixSort
     using CountType   = typename Traits::CountType;
     using KeyBits     = typename Traits::KeyBits;
 
+    static constexpr size_t INSERT_SORT_THRESHOLD = 64;
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: Missing comment (what it is; the motivation for constant: how it was derived)
PS. It is names "insertion sort", not "insert sort".

File: dbms/src/Common/RadixSort.h [hunk 4]
```diff
@@ -179,8 +198,101 @@ struct RadixSort
     static KeyBits keyToBits(Key x) { return ext::bit_cast<KeyBits>(x); }
     static Key bitsToKey(KeyBits x) { return ext::bit_cast<Key>(x); }
 
+    static inline void insertSortInternal(Element * arr, size_t size)
+    {
+        for (Element * i = arr + 1; i < arr + size; ++i)
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: Need to check if `arr + size` is calculated in every loop iteration (it is possible due to aliasing).

Comment (alexey-milovidov) [reply dbms/src/Common/RadixSort.h]: We must check if `std::sort` or `pdqsort` will be better.

File: dbms/src/Common/RadixSort.h [hunk 5]
```diff
@@ -179,8 +198,101 @@ struct RadixSort
     static KeyBits keyToBits(Key x) { return ext::bit_cast<KeyBits>(x); }
     static Key bitsToKey(KeyBits x) { return ext::bit_cast<Key>(x); }
 
+    static inline void insertSortInternal(Element * arr, size_t size)
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: No motivation for `inline`.

File: dbms/src/Common/RadixSort.h [hunk 6]
```diff
@@ -179,8 +198,101 @@ struct RadixSort
...
+                    *j = *(j - 1);
+                *j = tmp;
+            }
+        }
+    }
+
+    template <int PASS>
+    static inline void msdRadixSortInternal(Element * arr, size_t size, size_t limit)
+    {
+        Element *last_[HISTOGRAM_SIZE + 1];
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: Style.

File: dbms/src/Common/RadixSort.h [hunk 7]
```diff
@@ -179,8 +198,101 @@ struct RadixSort
...
+        Element ** last = last_ + 1;
+        size_t count[HISTOGRAM_SIZE] = {0};
+
+        for (Element * i = arr; i < arr + size; ++i)
+            ++count[getPart(PASS, *i)];
+
+        last_[0] = last_[1] = arr;
+
+        size_t bucketsForRecursion = HISTOGRAM_SIZE;
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: Style.

File: dbms/src/Common/RadixSort.h [hunk 8]
```diff
@@ -179,8 +198,101 @@ struct RadixSort
...
+                *j = tmp;
+            }
+        }
+    }
+
+    template <int PASS>
+    static inline void msdRadixSortInternal(Element * arr, size_t size, size_t limit)
+    {
+        Element *last_[HISTOGRAM_SIZE + 1];
+        Element ** last = last_ + 1;
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: Missing comment. What is going on here?

File: dbms/src/Common/RadixSort.h [hunk 9]
```diff
@@ -179,8 +198,101 @@ struct RadixSort
...
+                Element tmp = *i;
+                *i = *(i - 1);
+                for (j = i - 1; j > arr && Traits::compare(tmp, *(j - 1)); --j)
+                    *j = *(j - 1);
+                *j = tmp;
+            }
+        }
+    }
+
+    template <int PASS>
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: int?

File: dbms/src/Common/RadixSort.h [hunk 10]
```diff
@@ -179,8 +198,101 @@ struct RadixSort
...
+    static inline void msdRadixSortInternal(Element * arr, size_t size, size_t limit)
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: Missing comment.

File: dbms/src/Common/RadixSort.h (hunk 8 above)
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: Terrible naming.

File: dbms/src/Common/RadixSort.h [hunk 11]
```diff
@@ -179,8 +198,101 @@ struct RadixSort
...
+        if (size <= INSERT_SORT_THRESHOLD)
+            insertSortInternal(arr, size);
+        else
+            msdRadixSortInternal<PASS>(arr, size, limit);
+    }
+
 public:
-    static void execute(Element * arr, size_t size)
+    /* Least significant digit radix sort
+     * The most efficient stable general-purpose sorting algorithm
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: This statement is unfounded.

File: dbms/src/Common/RadixSort.h [hunk 12]
```diff
@@ -247,14 +359,30 @@ struct RadixSort
 
         allocator.deallocate(swap_buffer, size * sizeof(Element));
     }
+
+    /* Most significant digit radix sort
+     * Usually slower than LSD and is not stable, but allows partial sorting
+     * Based on https://github.com/voutcn/kxsort
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: You have partially copied some implementation details, but it requires to mention copyright and license.

File: dbms/src/Common/RadixSort.h (hunk 2 above)
Comment (kvinty) [reply dbms/src/Common/RadixSort.h]: Yes, that's a mistake.

File: dbms/src/Common/RadixSort.h (hunk 3 above)
Comment (kvinty) [reply dbms/src/Common/RadixSort.h]: Added a comment.
This constant is used in kxsort. I did not attempt to optimize it.

File: dbms/src/Common/RadixSort.h (hunk 4 above)
Comment (kvinty) [reply dbms/src/Common/RadixSort.h]: Created a pre-computed variable to ensure this never happens.

File: dbms/src/Common/RadixSort.h (hunk 5 above)
Comment (kvinty) [reply dbms/src/Common/RadixSort.h]: Removed the specifier. It does not affect performance.

**Pull request merged**